src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -888,70 +1003,186 @@ scaleCreditsForMerge SingleRun{} _ = ScaledCredits 0
 -- the sizes of the input runs. As as result, merge work would/could be more
 -- evenly distributed over time when the resident run is smaller than the worst
 -- case.
-scaleCreditsForMerge (MergingRun LevelLevelling _ _) c =
+scaleCreditsForMerge (MergingRun LevelLevelling _ _ _) (Credit c) =
     ScaledCredits (c * (1 + 4))
 -- A tiering merge has 5 runs at most (one could be held back to merged again)
 -- and must be completed before the level is full (once 4 more runs come in).
-scaleCreditsForMerge (MergingRun LevelTiering (NumRuns n) _) c =
+scaleCreditsForMerge (MergingRun LevelTiering (NumRuns n) _ _) (Credit c) =
     ScaledCredits ((c * n + 3) `div` 4)
     -- same as division rounding up: ceiling (c * n / 4)
 
-{-# SPECIALISE supplyMergeCredits :: ScaledCredits -> MergingRun IO h -> IO () #-}
--- TODO: implement doing merge werk in batches, instead of always taking the
--- MVar. The thresholds for doing merge work should be different for each level,
--- maybe co-prime?
+{-# SPECIALISE supplyMergeCredits :: ScaledCredits -> CreditThreshold -> MergingRun IO h -> IO () #-}
+-- | Supply the given amount of credits to a merging run. This /may/ cause an
+-- ongoing merge to progress.
+--
+-- See Note [Credits]
 supplyMergeCredits ::
-     (MonadSTM m, MonadST m, MonadMVar m, MonadMask m, MonadFix m)
+     forall m h. (MonadSTM m, MonadST m, MonadMVar m, MonadMask m, MonadFix m)
   => ScaledCredits
+  -> CreditThreshold
   -> MergingRun m h
   -> m ()
-supplyMergeCredits _ SingleRun{} = pure ()
-supplyMergeCredits (ScaledCredits c) (MergingRun _ _ var) = do
-    mergeIsDone <- withMVar var $ \case
+supplyMergeCredits _ _ SingleRun{} = pure ()
+supplyMergeCredits
+  (ScaledCredits c)
+  thresh@(CreditThreshold creditsThresh)
+  (MergingRun _ _ unspentCreditsVar mergeVar) = do
+    -- Add credits to the total of unspent credits
+    unspentCredits <- fetchAddInt unspentCreditsVar c
+    -- Compute our /estimate/ of what the new total of unspent credits is. The
+    -- /actual/ total might have been changed again by a different thread, but
+    -- that is okay because we will find out if\/when we run takeCredits.
+    let unspentCredits' = unspentCredits + c
+
+    -- Try to do some merging work if the new total now goes over the threshold.
+    --
+    -- TODO: when a merge is already completed, then could we skip most of the
+    -- stuff below?
```

Comment:
Finding that out would require reading an `MVar` every time (not just when going over the threshold), which sounds like it might end up being more expensive than credit accounting it might save. And in the common case when the merge is still ongoing it doesn't even save anything.

Still, it might be worth checking the performance impact, because reading the `mergeVar` would also allow us to to move `unspentCreditsVar` to an IMO more appropriate place and avoid the case where `expectCompletedMerge` still needs to do some merging (see other comments).

Reply:
> Finding that out would require reading an MVar every time (not just when going over the threshold), which sounds like it might end up being more expensive than credit accounting it might save. And in the common case when the merge is still ongoing it doesn't even save anything.

That's not the only way to do it. You could add an `IORef Bool` or something similar that is `False` initially and only gets updated to `True` once the merge is completed. It would be a single read + a single check every time, instead of doing credit accounting.

> Still, it might be worth checking the performance impact, because reading the mergeVar would also allow us to to move unspentCreditsVar to an IMO more appropriate place and avoid the case where expectCompletedMerge still needs to do some merging (see other comments).

As mentioned before, I don't think that is the way to go. The fact that we have `unspentCreditsVar` outside of the `MVar` is exactly so we can have a (largely) lock-free design

Reply:
Makes sense, although adding redundant state would need to be justified by a noticeable performance improvement, which I don't think we'll see. So I'd probably leave it as a TODO to try later.

Reply:
I've elaborated the TODO a bit